dfc/tests/integration_test.go

```diff
@@ -0,0 +1,301 @@
+	close(filenameCh)
+
+	for f := range filenameCh {
+		m.repFilenameCh <- repFile{repetitions: m.numGetsEachFile, filename: f}
+	}
+	m.wg.Wait()
+
+	// Step 4
+	m.wg.Add(m.num*m.numGetsEachFile + 2)
+	go func() {
+		// Failback is timed to ensure it doesn't finish before GETs begin
+		time.Sleep(17 * time.Second)
+		primarySetToOriginal(t)
+		m.wg.Done()
+	}()
+
+	go doRepeatedGets(&m)
+	doReregisterTarget(&m)
+
+	m.wg.Wait()
+	tlogf("%d GET errors before target's local bucket map was updated (expected)\n", m.numGetErrsBefore)
+	if m.numGetErrsAfter > maxNumGetErrs {
+		t.Fatalf("Found %d GET errors (should be no more than %d) after target's local bucket map was updated",
+			m.numGetErrsAfter, maxNumGetErrs)
+	} else {
+		tlogf("Found %d GET errors (should be no more than %d) after target's local bucket map was updated",
+			m.numGetErrsAfter, maxNumGetErrs)
+	}
+}
+
+func doRepeatedGets(m *metadata) {
+	for i := 0; i < 10; i++ {
+		m.semaphore <- struct{}{}
+	}
+	tlogf("Getting each of the %d files %d times from bucket %s...\n", m.num, m.numGetsEachFile, m.bucket)
+
+	// GET is timed so a large portion of requests will happen both before and after the target is re-registered
+	time.Sleep(m.delay)
+	for i := 0; i < m.num*m.numGetsEachFile; i++ {
+		go func() {
+			<-m.semaphore
+			defer func() {
+				m.semaphore <- struct{}{}
+				m.wg.Done()
+			}()
+
+			repFile := <-m.repFilenameCh
+			if repFile.repetitions > 0 {
+				repFile.repetitions -= 1
+				m.repFilenameCh <- repFile
+			}
+			_, _, err := client.Get(proxyurl, m.bucket, SmokeStr+"/"+repFile.filename, nil, nil, false, false)
+			if err != nil {
+				r := atomic.LoadUint64(&(m.reregistered))
+				if r == 1 {
+					atomic.AddUint64(&(m.numGetErrsAfter), 1)
+				} else {
+					atomic.AddUint64(&(m.numGetErrsBefore), 1)
+				}
+			}
+			g := atomic.AddUint64(&(m.getsCompleted), 1)
+			if g%5000 == 0 {
+				tlogf(" %d/%d GET requests completed...\n", g, m.num*m.numGetsEachFile)
+			}
+		}()
+	}
+}
+
+func doReregisterTarget(m *metadata) {
+	err := client.RegisterTarget(m.sid, &(m.smap))
+	checkFatal(err, m.t)
+PollLoop:
+	for i := 0; i < 25; i++ {
+		time.Sleep(time.Second)
+		m.smap = getClusterMap(httpclient, m.t)
+		lbNames, err := client.GetLocalBucketNames(m.targetDirectURL)
+		checkFatal(err, m.t)
+
+		if len(m.smap.Tmap) == m.origNumTargets {
+			for _, b := range lbNames.Local {
+				if b == m.bucket {
+					break PollLoop
+				}
+			}
+		}
+	}
+
+	if len(m.smap.Tmap) != m.origNumTargets {
+		m.t.Fatalf("Re-registration timed out: target %s, num targets now: %d, orig num targets: %d\n",
+			m.sid, len(m.smap.Tmap), m.origNumTargets)
+	}
+	tlogf("Re-registered target %s: the cluster is now back to %d targets\n", m.sid, m.origNumTargets)
+	s := atomic.CompareAndSwapUint64(&m.reregistered, 0, 1)
+	if !s {
+		m.t.Errorf("reregistered should have swapped from 0 to 1. Actual reregistered = %d\n", m.reregistered)
+	}
+	m.wg.Done()
+}
+
+func checkFatal(err error, t *testing.T) {
```

Comment:
I am not sure. On the one hand, it saves 2 lines every time. From the other hand, it hides the place of a fail - you always see in logs: "Failed at line 299" that is kind of useless. For small tests it is OK, for big ones it is not that good. 
Another way: it is possible to get the callers info(it is done in DFC) and tlogf the line which raises the error(httpcommon.go:Write)